src/bootstrap/builder.rs

```diff
@@ -404,6 +408,25 @@ impl<'a> Builder<'a> {
             parent: Cell::new(None),
         };
 
+        // Get paths to the test suites
+        let mut suites: Vec<PathBuf> = vec![];
+        if kind == Kind::Test{
+            for desc in Builder::get_step_descriptions(kind) {
+                let suite = (desc.should_run)(ShouldRun::new(&builder))
+                    .suite_path.unwrap_or("not set".to_string());
+                for p in paths {
+                    if p.starts_with(&suite) {
+                        suites.push(PathBuf::from(suite));
+                        break
```

Comment:
This feels wrong to me; if `./x.py test src/test/run-pass src/test/compile-fail` is invoked I'd expect the break to exit too early for us to populate suites for both tests.

It'd would also somewhat surprise me if this will work as-is when given a path to say `src/libstd` since those are all set as individual paths inside the pathset, whereas the "suite" detection only matches the last path given. It might be best to restructure to have suite-matching be separate and orthogonal to the path-matching to avoid that problem.

Reply:
Hmm, Could you please clarify on this if you don't mind.

Reply:
Basically, the way this code handles suites conflates them with the paths that could invoke a given Step. I think that's the wrong distinction to have as it adds little benefit and increases maintenance burden by connecting two concepts that are mostly unrelated. Instead, I would rather see these modifications contained to changes in `maybe_run` and [the code that calls it](https://github.com/rust-lang/rust/blob/49317cd511fbb60178bd5122e484609568938468/src/bootstrap/builder.rs#L206-L209). Essentially, I'd expect another case added that calls `should_run.is_suite_path(path)` or something like and then maybe_run can be called with a `PathSet::Suite` (making `PathSet` an enum). Does that help clarify? I can maybe explain in further detail.

Sorry for the somewhat moving target here!

Reply:
It's alright. I now get it. Thanks! Will let you know if I get any more questions.